### DISCUSSION TITLE
NVMe support

### DISCUSSION THREAD

**Discussion by nbanb** [virt-manager/virt-manager#732] [General]
Hi
I m new to virt-manager and I didn't find the way to emulate NVME drive. 
Is there a way to create a custom device which would be an NVME emulated drive ?
Thanks 
Regards
nbanba

**Comment by jaspi72** [comment]
You can use nvme as:
Add new hardware, search your nvme from pci...
In boot hit f2 and change boot order ;)

**Reply by nbanb** [reply]
Hi
Thanks for answer
I think it would require PCIE passthru, or a sort of exclusive access on the NVMe controller (like Intel VT-x/VT-d: [vt-directed-io-spec.pdf](http://www.intel.com/content/dam/www/public/us/en/documents/product-specifications/vt-directed-io-spec.pdf) ), no ?

My requirement is to 'emulate' 100% of an NVMe drive, whatever could be the underlying storage the VM image reside on.
(In my case it's NVMe but those NVMe drives do not support user to create "new" namespaces which could have a dedicated namespaces PCIE address on PCIE root tree which can be passed as a PCIE device with a passthru address)

When talking about emulation, I want to say that the image file (which is the VM emulate NVMe drive) can reside on an SAS, SATA or NVMe underlying storage (or/and a nested virtIO controller), but the controller in the VM would be an NVMe controller even if the physical host had no NVMe drives...

It should be portable as emulation is => no need to have specific hardware (ie: physical NVMe drives on host) to start a VM with an NVMe virtual drive

Thanks 
Kind regards,
nbanba

**Comment by phrdina** [comment]
Hi, currently this is not possible. Seems like QEMU supports this but libvirt doesn't. In order to make this work it would have to be implemented into libvirt first.

**Reply by nbanb** [reply]
Hi

Thanks for answer.
My need is to be able to 'emulate' NVMe drives / controler on VM whatever the host storage is ...
Do you think I should open an "issue" or a feature request at libvirt ?
Would it help to go forward in this subject ?

kind regards
nbanba

**Reply by phrdina** [reply]
Yes please, create libivrt issue https://gitlab.com/libvirt/libvirt and also virt-manager issue so we can track both.

**Comment by nbanb** [comment]
Dear @phrdina 

Sorry for the delay, it's done

To track this feature request these 2 issues are open :
[virt-manger #790 ](https://github.com/virt-manager/virt-manager/issues/790)
[libvirt #720](https://gitlab.com/libvirt/libvirt/-/issues/720)

Thanks
Kind regards
nbanba

**Comment by laomaiweng** [comment]
Hey, just an update on this, libvirt added support for emulated NVMe block devices with version 11.5.0 earlier this year: https://gitlab.com/libvirt/libvirt/-/blob/master/NEWS.rst?ref_type=heads&plain=1#L291

So it's now possible to add an emulated NVMe drive in virt-manager by directly editing the XML:
* add a VirtIO drive as usual
* maybe set a serial (the libvirt NEWS item sounds like this is required?)
* edit the drive's XML configuration:
  * set the target to NVMe: `<target dev='nvme0n1' bus='nvme'/>`
  * clear the `<address type="pci">` tag if there is one (NVMe emulated drives are hosted on an NVMe controller which libvirt will automatically add)

**Reply by roke-julian-lockhart** [reply]
@laomaiweng, <del>would be brilliant if they were set as the default storage type for at least Windows 11</del>, insofar as it's faster than VirtIO.

**Comment by phrdina** [comment]
NVMe support was added https://github.com/virt-manager/virt-manager/pull/1021 .
